## Importing libraries

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

## Data Preprocessing

In [2]:
df = pd.read_csv("/Users/kushal/Downloads/Crime_Incidents.csv", low_memory=False)

In [3]:
df2 = df[df['incident_type_primary'].isin({'SEXUAL ABUSE', 'RAPE', 'Other Sexual Offense',
                                 'Sexual Assault'})]

In [4]:
df3 = df2.reset_index(drop = True)
df3.head()

incident_id case_number       incident_datetime incident_type_primary  \
0    902969362  19-0250243  01/25/2015 12:00:00 PM          SEXUAL ABUSE   
1    902969363  19-0250244  01/25/2015 12:00:00 PM          SEXUAL ABUSE   
2    902969364  19-0250245  01/25/2018 12:00:00 PM          SEXUAL ABUSE   
3    902972631  19-0250450  01/25/2018 12:00:00 PM          SEXUAL ABUSE   
4    901019637  19-0120374  01/09/2019 12:00:00 PM          SEXUAL ABUSE   

                                incident_description  clearance_type  \
0  Buffalo Police are investigating this report o...             NaN   
1  Buffalo Police are investigating this report o...             NaN   
2  Buffalo Police are investigating this report o...             NaN   
3  Buffalo Police are investigating this report o...             NaN   
4  Buffalo Police are investigating this report o...             NaN   

                address_1  address_2     city state  ...  \
0     100 Block CULVER RD        NaN  BUFFALO    NY  ...   
1   1 Block VANDERBILT ST        NaN  BUFFALO    NY  ...   
2  200 Block HEMPSTEAD AV        NaN  BUFFALO    NY  ...   
3       300 Block 14TH ST        NaN  BUFFALO    NY  ...   
4    1 Block WAKEFIELD AV        NaN  BUFFALO    NY  ...   

                                       location hour_of_day  day_of_week  \
0  POINT (-78.81745109743173 42.83980951202153)          12       Sunday   
1    POINT (-78.805856858726 42.89218618798844)          12       Sunday   
2   POINT (-78.80506119711333 42.9304922315033)          12     Thursday   
3   POINT (-78.8906598853704 42.91089527708419)          12     Thursday   
4   POINT (-78.83565825907772 42.9342428518739)          12    Wednesday   

   parent_incident_type Council Districts Police Districts Zip Codes  Tracts  \
0  Other Sexual Offense               6.0              4.0       9.0    74.0   
1  Other Sexual Offense               2.0              3.0      17.0    28.0   
2  Other Sexual Offense               8.0              2.0      11.0    63.0   
3  Other Sexual Offense               3.0              1.0       2.0    41.0   
4  Other Sexual Offense               9.0              2.0      18.0    19.0   

  Block Groups  Blocks  
0         83.0   535.0  
1        142.0  1305.0  
2        126.0  1522.0  
3         96.0  3296.0  
4         13.0  1373.0  

[5 rows x 26 columns]

In [5]:
date_times = []
for num in range(df2.shape[0]):
    date_times.append(df2['incident_datetime'].iloc[num].split(" ", 2)[0])
date_times[:5]

['01/25/2015', '01/25/2015', '01/25/2018', '01/25/2018', '01/09/2019']

In [6]:
date_times_2 = pd.to_datetime(date_times,infer_datetime_format=True)
df3['year'] = date_times_2.year
df3['month'] = date_times_2.month
df3.head(5)

incident_id case_number       incident_datetime incident_type_primary  \
0    902969362  19-0250243  01/25/2015 12:00:00 PM          SEXUAL ABUSE   
1    902969363  19-0250244  01/25/2015 12:00:00 PM          SEXUAL ABUSE   
2    902969364  19-0250245  01/25/2018 12:00:00 PM          SEXUAL ABUSE   
3    902972631  19-0250450  01/25/2018 12:00:00 PM          SEXUAL ABUSE   
4    901019637  19-0120374  01/09/2019 12:00:00 PM          SEXUAL ABUSE   

                                incident_description  clearance_type  \
0  Buffalo Police are investigating this report o...             NaN   
1  Buffalo Police are investigating this report o...             NaN   
2  Buffalo Police are investigating this report o...             NaN   
3  Buffalo Police are investigating this report o...             NaN   
4  Buffalo Police are investigating this report o...             NaN   

                address_1  address_2     city state  ... day_of_week  \
0     100 Block CULVER RD        NaN  BUFFALO    NY  ...      Sunday   
1   1 Block VANDERBILT ST        NaN  BUFFALO    NY  ...      Sunday   
2  200 Block HEMPSTEAD AV        NaN  BUFFALO    NY  ...    Thursday   
3       300 Block 14TH ST        NaN  BUFFALO    NY  ...    Thursday   
4    1 Block WAKEFIELD AV        NaN  BUFFALO    NY  ...   Wednesday   

   parent_incident_type  Council Districts  Police Districts Zip Codes Tracts  \
0  Other Sexual Offense                6.0               4.0       9.0   74.0   
1  Other Sexual Offense                2.0               3.0      17.0   28.0   
2  Other Sexual Offense                8.0               2.0      11.0   63.0   
3  Other Sexual Offense                3.0               1.0       2.0   41.0   
4  Other Sexual Offense                9.0               2.0      18.0   19.0   

  Block Groups  Blocks  year month  
0         83.0   535.0  2015     1  
1        142.0  1305.0  2015     1  
2        126.0  1522.0  2018     1  
3         96.0  3296.0  2018     1  
4         13.0  1373.0  2019     1  

[5 rows x 28 columns]

In [7]:
df4 = df[df['incident_type_primary'].isin({'ASSAULT', 'LARCENY/THEFT',  'UUV', 'BURGLARY',
        'ROBBERY', 'CRIM NEGLIGENT HOMICIDE', 'THEFT OF SERVICES',
       'MURDER', 'Breaking & Entering', 'Assault', 'Theft', 'Robbery',
       'Theft of Vehicle', 'AGG ASSAULT ON P/OFFICER',
         'Homicide',
       'AGGR ASSAULT', 'MANSLAUGHTER'})]
df4.head(5)

incident_id case_number       incident_datetime incident_type_primary  \
0    707535288  15-0650868  03/06/2015 07:35:15 AM               ASSAULT   
1    901476393  19-0160417  01/04/2019 12:00:00 PM         LARCENY/THEFT   
2    902552636  19-0220443  01/04/2019 12:00:00 PM         LARCENY/THEFT   
8    854542298  18-1380353  03/01/2018 12:00:00 PM         LARCENY/THEFT   
9    894581032  18-3510529  11/20/2018 12:00:00 PM         LARCENY/THEFT   

                                incident_description  clearance_type  \
0                                            ASSAULT             NaN   
1  Buffalo Police are investigating this report o...             NaN   
2  Buffalo Police are investigating this report o...             NaN   
8  Buffalo Police are investigating this report o...             NaN   
9  Buffalo Police are investigating this report o...             NaN   

               address_1  address_2     city state  ...  \
0         HOPKINS @ GOOD        NaN  BUFFALO    NY  ...   
1      400 Block GOLD ST        NaN  BUFFALO    NY  ...   
2  1000 Block GENESEE ST        NaN  BUFFALO    NY  ...   
8    300 Block WALDEN AV        NaN  BUFFALO    NY  ...   
9    200 Block HUDSON ST        NaN  BUFFALO    NY  ...   

                                       location hour_of_day  day_of_week  \
0      POINT (-78.818186333039 42.846887288511)           7       Friday   
1  POINT (-78.80519990642684 42.89630328831633)          12       Friday   
2  POINT (-78.8442581986752 42.901489184581514)          12       Friday   
8  POINT (-78.82455864379337 42.90326103136154)          12     Thursday   
9  POINT (-78.87146383534977 42.89289047689623)          12      Tuesday   

   parent_incident_type Council Districts Police Districts Zip Codes  Tracts  \
0               Assault               6.0              4.0       9.0    71.0   
1                 Theft               2.0              3.0      17.0    28.0   
2                 Theft               5.0              3.0       4.0    26.0   
8                 Theft               5.0              3.0       4.0    66.0   
9                 Theft               4.0              1.0      14.0    77.0   

  Block Groups  Blocks  
0        211.0   126.0  
1         24.0   314.0  
2          8.0  3027.0  
8        242.0  2301.0  
9         31.0   914.0  

[5 rows x 26 columns]

## Latitude and longitude computation

In [8]:
df_lat_long = df3.loc[np.where(df3['location'].isna() == False)[0],].reset_index()

In [9]:
df_lat_long.shape
df_lat_long.tail()

index  incident_id case_number       incident_datetime  \
4699   4718    940771784  19-2490644  09/06/2019 04:00:00 PM   
4700   4719    940988520  19-2510719  09/07/2019 09:30:00 PM   
4701   4720    940883616  19-2510153  09/08/2019 02:00:00 AM   
4702   4721    940988519  19-2510937  09/08/2019 11:06:36 PM   
4703   4722    942190728  19-2620791  09/19/2019 06:40:00 PM   

     incident_type_primary                               incident_description  \
4699          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4700          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4701          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4702                  RAPE  Buffalo Police are investigating this report o...   
4703                  RAPE  Buffalo Police are investigating this report o...   

      clearance_type                    address_1  address_2     city  ...  \
4699             NaN         1300 Block BAILEY AV        NaN  BUFFALO  ...   
4700             NaN          400 Block HEWITT AV        NaN  BUFFALO  ...   
4701             NaN            300 Block 14TH ST        NaN  BUFFALO  ...   
4702             NaN        700 Block FILLMORE AV        NaN  BUFFALO  ...   
4703             NaN  RHODE ISLAND ST & NORMAL AV        NaN  BUFFALO  ...   

     day_of_week  parent_incident_type Council Districts  Police Districts  \
4699      Friday  Other Sexual Offense               1.0               5.0   
4700    Saturday  Other Sexual Offense               8.0               2.0   
4701      Sunday  Other Sexual Offense               3.0               1.0   
4702      Sunday        Sexual Assault               5.0               3.0   
4703    Thursday        Sexual Assault               3.0               1.0   

      Zip Codes Tracts Block Groups  Blocks  year month  
4699       10.0    9.0         91.0  3273.0  2019     9  
4700       11.0    7.0        155.0  2134.0  2019     9  
4701        2.0   41.0         96.0  1714.0  2019     9  
4702        8.0   62.0         57.0  2160.0  2019     9  
4703        2.0   41.0        265.0   328.0  2019     9  

[5 rows x 29 columns]

In [10]:
temp = df_lat_long['location'].str.split(" ")
df_lat_long['Longitude'] = [float(temp[x][1][1:]) for x in range(temp.shape[0])]
df_lat_long['Latitude'] = [float(temp[x][2][:-1]) for x in range(temp.shape[0])]

In [11]:
temp = df_lat_long['location'].str.split(" ")
temp.head(3)

0    [POINT, (-78.81745109743173, 42.83980951202153)]
1      [POINT, (-78.805856858726, 42.89218618798844)]
2     [POINT, (-78.80506119711333, 42.9304922315033)]
Name: location, dtype: object

In [12]:
df_with_latlong = df_lat_long
df_with_latlong.head()

index  incident_id case_number       incident_datetime  \
0      0    902969362  19-0250243  01/25/2015 12:00:00 PM   
1      1    902969363  19-0250244  01/25/2015 12:00:00 PM   
2      2    902969364  19-0250245  01/25/2018 12:00:00 PM   
3      3    902972631  19-0250450  01/25/2018 12:00:00 PM   
4      4    901019637  19-0120374  01/09/2019 12:00:00 PM   

  incident_type_primary                               incident_description  \
0          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
1          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
2          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
3          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4          SEXUAL ABUSE  Buffalo Police are investigating this report o...   

   clearance_type               address_1  address_2     city  ...  \
0             NaN     100 Block CULVER RD        NaN  BUFFALO  ...   
1             NaN   1 Block VANDERBILT ST        NaN  BUFFALO  ...   
2             NaN  200 Block HEMPSTEAD AV        NaN  BUFFALO  ...   
3             NaN       300 Block 14TH ST        NaN  BUFFALO  ...   
4             NaN    1 Block WAKEFIELD AV        NaN  BUFFALO  ...   

  Council Districts Police Districts Zip Codes  Tracts  Block Groups  Blocks  \
0               6.0              4.0       9.0    74.0          83.0   535.0   
1               2.0              3.0      17.0    28.0         142.0  1305.0   
2               8.0              2.0      11.0    63.0         126.0  1522.0   
3               3.0              1.0       2.0    41.0          96.0  3296.0   
4               9.0              2.0      18.0    19.0          13.0  1373.0   

   year month  Longitude   Latitude  
0  2015     1 -78.817451  42.839810  
1  2015     1 -78.805857  42.892186  
2  2018     1 -78.805061  42.930492  
3  2018     1 -78.890660  42.910895  
4  2019     1 -78.835658  42.934243  

[5 rows x 31 columns]

## University at buffalo south campus region

In [13]:
df_with_latlong_filtered1_sexual= df_with_latlong.loc[np.where((df_with_latlong['Latitude'] >= 42.948) & (df_with_latlong['Latitude'] <= 42.958) & 
         (df_with_latlong['Longitude'] <= -78.7815) &
        (df_with_latlong['Longitude'] >= -78.825))]

In [39]:
df_with_latlong_filtered1_sexual.shape

(33, 31)

In [38]:
df_with_latlong_filtered1_sexual.head()

index  incident_id case_number       incident_datetime  \
446    465    656416454  15-0070868  01/07/2015 07:30:00 AM   
650    669    826998998  17-3220793  11/18/2017 06:00:00 AM   
661    680    775993057  16-2830316  05/20/2016 12:00:00 PM   
832    851     63773686  11-1180213  04/28/2011 01:00:00 AM   
964    983     85157479  11-2970264  10/22/2011 09:00:00 PM   

    incident_type_primary                               incident_description  \
446          SEXUAL ABUSE                                       SEXUAL ABUSE   
650          SEXUAL ABUSE                                       SEXUAL ABUSE   
661          SEXUAL ABUSE                                       SEXUAL ABUSE   
832          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
964                  RAPE  Buffalo Police are investigating this report o...   

     clearance_type              address_1  address_2     city  ...  \
446             NaN  300 Block HIGHGATE AV        NaN  BUFFALO  ...   
650             NaN  300 Block WINSPEAR AV        NaN  BUFFALO  ...   
661             NaN  400 Block WINSPEAR AV        NaN  BUFFALO  ...   
832             NaN  1 Block KENSINGTON AV        NaN  BUFFALO  ...   
964             NaN               LEROY AV        NaN  BUFFALO  ...   

    Council Districts Police Districts Zip Codes  Tracts  Block Groups  \
446               8.0              2.0      11.0    56.0         257.0   
650               8.0              2.0      11.0     7.0         131.0   
661               8.0              2.0      11.0     7.0         131.0   
832               8.0              2.0       6.0     8.0         149.0   
964               8.0              2.0       6.0     8.0         149.0   

     Blocks  year month  Longitude   Latitude  
446   333.0  2015     1 -78.814276  42.948254  
650  1968.0  2017    11 -78.812261  42.948725  
661  1929.0  2016     5 -78.808787  42.948677  
832  3216.0  2011     4 -78.817557  42.957975  
964  3216.0  2011    10 -78.817557  42.957975  

[5 rows x 31 columns]

## Airport Region

In [15]:
df_with_latlong_filtered2_sexual= df_with_latlong.loc[np.where((df_with_latlong['Latitude'] >= 42.875) & (df_with_latlong['Latitude'] <= 42.885) & 
         (df_with_latlong['Longitude'] <= -78.873) &
        (df_with_latlong['Longitude'] >= -78.883))]

In [41]:
df_with_latlong_filtered2_sexual.shape

(130, 31)

In [40]:
df_with_latlong_filtered2_sexual.head()

index  incident_id case_number       incident_datetime  \
28      28    901653246  19-0160579  01/16/2019 12:00:00 PM   
56      56    889733513  18-3200278  11/16/2018 09:44:06 AM   
93      93    894551049  18-3510309  12/17/2018 11:48:00 AM   
148    148    882312777  18-2940481  10/21/2018 01:30:00 AM   
173    173    900375629  19-0090418  01/10/2014 12:00:00 PM   

    incident_type_primary                               incident_description  \
28           SEXUAL ABUSE  Buffalo Police are investigating this report o...   
56                   RAPE  Buffalo Police are investigating this report o...   
93           SEXUAL ABUSE  Buffalo Police are investigating this report o...   
148                  RAPE  Buffalo Police are investigating this report o...   
173          SEXUAL ABUSE  Buffalo Police are investigating this report o...   

     clearance_type            address_1  address_2     city  ...  \
28              NaN  1 Block FRANKLIN ST        NaN  BUFFALO  ...   
56              NaN   1 Block NIAGARA PA        NaN  BUFFALO  ...   
93              NaN    1 Block MARINE DR        NaN  BUFFALO  ...   
148             NaN  200 Block MARINE DR        NaN  BUFFALO  ...   
173             NaN  1 Block FRANKLIN ST        NaN  BUFFALO  ...   

    Council Districts Police Districts Zip Codes  Tracts  Block Groups  \
28                5.0              1.0       3.0    77.0          31.0   
56                4.0              1.0       3.0    77.0          31.0   
93                5.0              1.0       3.0    67.0         203.0   
148               5.0              1.0       3.0    67.0         203.0   
173               5.0              1.0       3.0    77.0          31.0   

     Blocks  year month  Longitude   Latitude  
28    270.0  2019     1 -78.878023  42.883160  
56   2598.0  2018    11 -78.876501  42.884920  
93   2859.0  2018    12 -78.881413  42.879091  
148  1203.0  2018    10 -78.879945  42.879424  
173   270.0  2014     1 -78.878023  42.883160  

[5 rows x 31 columns]

##  University at Buffalo North Campus Region

In [17]:
df_with_latlong_filtered3_sexual= df_with_latlong.loc[np.where((df_with_latlong['Latitude'] >= 42.975) & (df_with_latlong['Latitude'] <= 43.025) & 
         (df_with_latlong['Longitude'] <= -78.765) &
        (df_with_latlong['Longitude'] >= -78.815))]

In [43]:
df_with_latlong_filtered3_sexual.shape

(0, 31)

In [42]:
df_with_latlong_filtered3_sexual.head()

Empty DataFrame
Columns: [index, incident_id, case_number, incident_datetime, incident_type_primary, incident_description, clearance_type, address_1, address_2, city, state, zip, country, latitude, longitude, created_at, updated_at, location, hour_of_day, day_of_week, parent_incident_type, Council Districts, Police Districts, Zip Codes, Tracts, Block Groups, Blocks, year, month, Longitude, Latitude]
Index: []

[0 rows x 31 columns]

## Downtown Region Buffalo


In [19]:
df_with_latlong_filtered4_sexual= df_with_latlong.loc[np.where((df_with_latlong['Latitude'] >= 42.881) & (df_with_latlong['Latitude'] <= 42.891) & 
         (df_with_latlong['Longitude'] <= -78.870) &
        (df_with_latlong['Longitude'] >= -78.880))]

In [20]:
df_with_latlong_filtered4_sexual.shape

(227, 31)

In [21]:
df_with_latlong_filtered4_sexual.head()

index  incident_id case_number       incident_datetime  \
28      28    901653246  19-0160579  01/16/2019 12:00:00 PM   
56      56    889733513  18-3200278  11/16/2018 09:44:06 AM   
68      68    856495489  18-1560382  05/29/2018 06:40:00 AM   
115    115    914894939  19-1140651  04/24/2019 12:00:00 PM   
173    173    900375629  19-0090418  01/10/2014 12:00:00 PM   

    incident_type_primary                               incident_description  \
28           SEXUAL ABUSE  Buffalo Police are investigating this report o...   
56                   RAPE  Buffalo Police are investigating this report o...   
68                   RAPE  Buffalo Police are investigating this report o...   
115          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
173          SEXUAL ABUSE  Buffalo Police are investigating this report o...   

     clearance_type            address_1  address_2     city  ...  \
28              NaN  1 Block FRANKLIN ST        NaN  BUFFALO  ...   
56              NaN   1 Block NIAGARA PA        NaN  BUFFALO  ...   
68              NaN    600 Block MAIN ST        NaN  BUFFALO  ...   
115             NaN     1 Block COURT ST        NaN  BUFFALO  ...   
173             NaN  1 Block FRANKLIN ST        NaN  BUFFALO  ...   

    Council Districts Police Districts Zip Codes  Tracts  Block Groups  \
28                5.0              1.0       3.0    77.0          31.0   
56                4.0              1.0       3.0    77.0          31.0   
68                4.0              1.0      14.0    77.0          31.0   
115               4.0              1.0       3.0    77.0          31.0   
173               5.0              1.0       3.0    77.0          31.0   

     Blocks  year month  Longitude   Latitude  
28    270.0  2019     1 -78.878023  42.883160  
56   2598.0  2018    11 -78.876501  42.884920  
68    914.0  2018     5 -78.872455  42.890176  
115  2313.0  2019     4 -78.877103  42.885852  
173   270.0  2014     1 -78.878023  42.883160  

[5 rows x 31 columns]

Here we observed downtown region in buffalo has the maximum number of sexual crimes. Second highest region is the airport region. Next is the south campus region where sexual crimes rate is average. North campus is the safest place observed.

In [22]:
df_with_latlong.head()

index  incident_id case_number       incident_datetime  \
0      0    902969362  19-0250243  01/25/2015 12:00:00 PM   
1      1    902969363  19-0250244  01/25/2015 12:00:00 PM   
2      2    902969364  19-0250245  01/25/2018 12:00:00 PM   
3      3    902972631  19-0250450  01/25/2018 12:00:00 PM   
4      4    901019637  19-0120374  01/09/2019 12:00:00 PM   

  incident_type_primary                               incident_description  \
0          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
1          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
2          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
3          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4          SEXUAL ABUSE  Buffalo Police are investigating this report o...   

   clearance_type               address_1  address_2     city  ...  \
0             NaN     100 Block CULVER RD        NaN  BUFFALO  ...   
1             NaN   1 Block VANDERBILT ST        NaN  BUFFALO  ...   
2             NaN  200 Block HEMPSTEAD AV        NaN  BUFFALO  ...   
3             NaN       300 Block 14TH ST        NaN  BUFFALO  ...   
4             NaN    1 Block WAKEFIELD AV        NaN  BUFFALO  ...   

  Council Districts Police Districts Zip Codes  Tracts  Block Groups  Blocks  \
0               6.0              4.0       9.0    74.0          83.0   535.0   
1               2.0              3.0      17.0    28.0         142.0  1305.0   
2               8.0              2.0      11.0    63.0         126.0  1522.0   
3               3.0              1.0       2.0    41.0          96.0  3296.0   
4               9.0              2.0      18.0    19.0          13.0  1373.0   

   year month  Longitude   Latitude  
0  2015     1 -78.817451  42.839810  
1  2015     1 -78.805857  42.892186  
2  2018     1 -78.805061  42.930492  
3  2018     1 -78.890660  42.910895  
4  2019     1 -78.835658  42.934243  

[5 rows x 31 columns]

In [23]:
df_with_latlong_new= df_with_latlong.loc[np.where(df_with_latlong['year'] > 2009)].reset_index(drop = True)


In [24]:
df_with_latlong_new.head()

index  incident_id case_number       incident_datetime  \
0      0    902969362  19-0250243  01/25/2015 12:00:00 PM   
1      1    902969363  19-0250244  01/25/2015 12:00:00 PM   
2      2    902969364  19-0250245  01/25/2018 12:00:00 PM   
3      3    902972631  19-0250450  01/25/2018 12:00:00 PM   
4      4    901019637  19-0120374  01/09/2019 12:00:00 PM   

  incident_type_primary                               incident_description  \
0          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
1          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
2          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
3          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4          SEXUAL ABUSE  Buffalo Police are investigating this report o...   

   clearance_type               address_1  address_2     city  ...  \
0             NaN     100 Block CULVER RD        NaN  BUFFALO  ...   
1             NaN   1 Block VANDERBILT ST        NaN  BUFFALO  ...   
2             NaN  200 Block HEMPSTEAD AV        NaN  BUFFALO  ...   
3             NaN       300 Block 14TH ST        NaN  BUFFALO  ...   
4             NaN    1 Block WAKEFIELD AV        NaN  BUFFALO  ...   

  Council Districts Police Districts Zip Codes  Tracts  Block Groups  Blocks  \
0               6.0              4.0       9.0    74.0          83.0   535.0   
1               2.0              3.0      17.0    28.0         142.0  1305.0   
2               8.0              2.0      11.0    63.0         126.0  1522.0   
3               3.0              1.0       2.0    41.0          96.0  3296.0   
4               9.0              2.0      18.0    19.0          13.0  1373.0   

   year month  Longitude   Latitude  
0  2015     1 -78.817451  42.839810  
1  2015     1 -78.805857  42.892186  
2  2018     1 -78.805061  42.930492  
3  2018     1 -78.890660  42.910895  
4  2019     1 -78.835658  42.934243  

[5 rows x 31 columns]

In [25]:
df_with_latlong_new.shape

(3154, 31)

In [26]:
df_with_latlong_new.head()

index  incident_id case_number       incident_datetime  \
0      0    902969362  19-0250243  01/25/2015 12:00:00 PM   
1      1    902969363  19-0250244  01/25/2015 12:00:00 PM   
2      2    902969364  19-0250245  01/25/2018 12:00:00 PM   
3      3    902972631  19-0250450  01/25/2018 12:00:00 PM   
4      4    901019637  19-0120374  01/09/2019 12:00:00 PM   

  incident_type_primary                               incident_description  \
0          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
1          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
2          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
3          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4          SEXUAL ABUSE  Buffalo Police are investigating this report o...   

   clearance_type               address_1  address_2     city  ...  \
0             NaN     100 Block CULVER RD        NaN  BUFFALO  ...   
1             NaN   1 Block VANDERBILT ST        NaN  BUFFALO  ...   
2             NaN  200 Block HEMPSTEAD AV        NaN  BUFFALO  ...   
3             NaN       300 Block 14TH ST        NaN  BUFFALO  ...   
4             NaN    1 Block WAKEFIELD AV        NaN  BUFFALO  ...   

  Council Districts Police Districts Zip Codes  Tracts  Block Groups  Blocks  \
0               6.0              4.0       9.0    74.0          83.0   535.0   
1               2.0              3.0      17.0    28.0         142.0  1305.0   
2               8.0              2.0      11.0    63.0         126.0  1522.0   
3               3.0              1.0       2.0    41.0          96.0  3296.0   
4               9.0              2.0      18.0    19.0          13.0  1373.0   

   year month  Longitude   Latitude  
0  2015     1 -78.817451  42.839810  
1  2015     1 -78.805857  42.892186  
2  2018     1 -78.805061  42.930492  
3  2018     1 -78.890660  42.910895  
4  2019     1 -78.835658  42.934243  

[5 rows x 31 columns]

In [27]:
df_with_latlong_new['AreaCode']= np.nan

In [28]:
a= np.where((df_with_latlong_new['Latitude'] >= 42.881) & (df_with_latlong_new['Latitude'] <= 42.891) & 
         (df_with_latlong_new['Longitude'] <= -78.870) &
        (df_with_latlong_new['Longitude'] >= -78.880))[0]

b= np.where((df_with_latlong_new['Latitude'] >= 42.875) & (df_with_latlong_new['Latitude'] <= 42.885) & 
         (df_with_latlong_new['Longitude'] <= -78.873) &
        (df_with_latlong_new['Longitude'] >= -78.883))[0]

c= np.where((df_with_latlong_new['Latitude'] >= 42.997) & (df_with_latlong_new['Latitude'] <=43.007 ) & 
         (df_with_latlong_new['Longitude'] <= -78.779) &
        (df_with_latlong_new['Longitude'] >= -78.789))[0]

d= np.where((df_with_latlong_new['Latitude'] >= 42.948) & (df_with_latlong_new['Latitude'] <=42.958 ) & 
         (df_with_latlong_new['Longitude'] <= -78.815) &
        (df_with_latlong_new['Longitude'] >= -78.825))[0]

In [29]:
df_with_latlong_new.loc[(a),'AreaCode']= 'Downtown'
df_with_latlong_new.loc[(b),'AreaCode']= 'Airport'
df_with_latlong_new.loc[(c),'AreaCode']= 'UB NorthCampus'
df_with_latlong_new.loc[(d),'AreaCode']= 'UB SouthCampus'

In [30]:
df_with_latlong_new.shape

(3154, 32)

In [31]:
df_with_latlong_new['AreaCode'].unique()

array([nan, 'Airport', 'Downtown', 'UB SouthCampus'], dtype=object)

In [32]:
df_with_latlong_new.head()

index  incident_id case_number       incident_datetime  \
0      0    902969362  19-0250243  01/25/2015 12:00:00 PM   
1      1    902969363  19-0250244  01/25/2015 12:00:00 PM   
2      2    902969364  19-0250245  01/25/2018 12:00:00 PM   
3      3    902972631  19-0250450  01/25/2018 12:00:00 PM   
4      4    901019637  19-0120374  01/09/2019 12:00:00 PM   

  incident_type_primary                               incident_description  \
0          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
1          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
2          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
3          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4          SEXUAL ABUSE  Buffalo Police are investigating this report o...   

   clearance_type               address_1  address_2     city  ...  \
0             NaN     100 Block CULVER RD        NaN  BUFFALO  ...   
1             NaN   1 Block VANDERBILT ST        NaN  BUFFALO  ...   
2             NaN  200 Block HEMPSTEAD AV        NaN  BUFFALO  ...   
3             NaN       300 Block 14TH ST        NaN  BUFFALO  ...   
4             NaN    1 Block WAKEFIELD AV        NaN  BUFFALO  ...   

  Police Districts Zip Codes Tracts  Block Groups  Blocks  year month  \
0              4.0       9.0   74.0          83.0   535.0  2015     1   
1              3.0      17.0   28.0         142.0  1305.0  2015     1   
2              2.0      11.0   63.0         126.0  1522.0  2018     1   
3              1.0       2.0   41.0          96.0  3296.0  2018     1   
4              2.0      18.0   19.0          13.0  1373.0  2019     1   

   Longitude   Latitude AreaCode  
0 -78.817451  42.839810      NaN  
1 -78.805857  42.892186      NaN  
2 -78.805061  42.930492      NaN  
3 -78.890660  42.910895      NaN  
4 -78.835658  42.934243      NaN  

[5 rows x 32 columns]

In [33]:
df_with_latlong_new['AreaCode'].fillna(0, inplace= True)
df_with_latlong_new.head()

index  incident_id case_number       incident_datetime  \
0      0    902969362  19-0250243  01/25/2015 12:00:00 PM   
1      1    902969363  19-0250244  01/25/2015 12:00:00 PM   
2      2    902969364  19-0250245  01/25/2018 12:00:00 PM   
3      3    902972631  19-0250450  01/25/2018 12:00:00 PM   
4      4    901019637  19-0120374  01/09/2019 12:00:00 PM   

  incident_type_primary                               incident_description  \
0          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
1          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
2          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
3          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4          SEXUAL ABUSE  Buffalo Police are investigating this report o...   

   clearance_type               address_1  address_2     city  ...  \
0             NaN     100 Block CULVER RD        NaN  BUFFALO  ...   
1             NaN   1 Block VANDERBILT ST        NaN  BUFFALO  ...   
2             NaN  200 Block HEMPSTEAD AV        NaN  BUFFALO  ...   
3             NaN       300 Block 14TH ST        NaN  BUFFALO  ...   
4             NaN    1 Block WAKEFIELD AV        NaN  BUFFALO  ...   

  Police Districts Zip Codes Tracts  Block Groups  Blocks  year month  \
0              4.0       9.0   74.0          83.0   535.0  2015     1   
1              3.0      17.0   28.0         142.0  1305.0  2015     1   
2              2.0      11.0   63.0         126.0  1522.0  2018     1   
3              1.0       2.0   41.0          96.0  3296.0  2018     1   
4              2.0      18.0   19.0          13.0  1373.0  2019     1   

   Longitude   Latitude AreaCode  
0 -78.817451  42.839810        0  
1 -78.805857  42.892186        0  
2 -78.805061  42.930492        0  
3 -78.890660  42.910895        0  
4 -78.835658  42.934243        0  

[5 rows x 32 columns]

In [34]:
df_with_latlong_new['AreaCode'].replace(0,'other',inplace=True)
df_with_latlong_new.head()

index  incident_id case_number       incident_datetime  \
0      0    902969362  19-0250243  01/25/2015 12:00:00 PM   
1      1    902969363  19-0250244  01/25/2015 12:00:00 PM   
2      2    902969364  19-0250245  01/25/2018 12:00:00 PM   
3      3    902972631  19-0250450  01/25/2018 12:00:00 PM   
4      4    901019637  19-0120374  01/09/2019 12:00:00 PM   

  incident_type_primary                               incident_description  \
0          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
1          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
2          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
3          SEXUAL ABUSE  Buffalo Police are investigating this report o...   
4          SEXUAL ABUSE  Buffalo Police are investigating this report o...   

   clearance_type               address_1  address_2     city  ...  \
0             NaN     100 Block CULVER RD        NaN  BUFFALO  ...   
1             NaN   1 Block VANDERBILT ST        NaN  BUFFALO  ...   
2             NaN  200 Block HEMPSTEAD AV        NaN  BUFFALO  ...   
3             NaN       300 Block 14TH ST        NaN  BUFFALO  ...   
4             NaN    1 Block WAKEFIELD AV        NaN  BUFFALO  ...   

  Police Districts Zip Codes Tracts  Block Groups  Blocks  year month  \
0              4.0       9.0   74.0          83.0   535.0  2015     1   
1              3.0      17.0   28.0         142.0  1305.0  2015     1   
2              2.0      11.0   63.0         126.0  1522.0  2018     1   
3              1.0       2.0   41.0          96.0  3296.0  2018     1   
4              2.0      18.0   19.0          13.0  1373.0  2019     1   

   Longitude   Latitude AreaCode  
0 -78.817451  42.839810    other  
1 -78.805857  42.892186    other  
2 -78.805061  42.930492    other  
3 -78.890660  42.910895    other  
4 -78.835658  42.934243    other  

[5 rows x 32 columns]

In [35]:
file = df_with_latlong_new[['incident_type_primary','year','AreaCode']] 
file['occurrence'] = 1

/Users/kushal/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [36]:
file.head()

incident_type_primary  year AreaCode  occurrence
0          SEXUAL ABUSE  2015    other           1
1          SEXUAL ABUSE  2015    other           1
2          SEXUAL ABUSE  2018    other           1
3          SEXUAL ABUSE  2018    other           1
4          SEXUAL ABUSE  2019    other           1

In [37]:
file2 = file.groupby(['year','AreaCode', 'incident_type_primary']).agg({'occurrence':'sum'}).reset_index()
file2.head

<bound method NDFrame.head of     year        AreaCode incident_type_primary  occurrence
0   2010         Airport                  RAPE          10
1   2010         Airport          SEXUAL ABUSE           4
2   2010        Downtown                  RAPE           6
3   2010        Downtown          SEXUAL ABUSE           2
4   2010  UB SouthCampus                  RAPE           3
5   2010  UB SouthCampus          SEXUAL ABUSE           3
6   2010           other                  RAPE         186
7   2010           other          SEXUAL ABUSE         145
8   2011         Airport                  RAPE           2
9   2011         Airport          SEXUAL ABUSE           2
10  2011        Downtown                  RAPE           2
11  2011        Downtown          SEXUAL ABUSE           4
12  2011  UB SouthCampus                  RAPE           1
13  2011  UB SouthCampus          SEXUAL ABUSE           1
14  2011           other                  RAPE         140
15  2011           other  

Thus after region analysis we can conclude that university at buffalo north campus is the most safest place for women
as there were no sexual crimes noticed in the past 10 years. Also, downtown is the most unsafest place for women as there were total 227 sexual crimes since the past 10 years and airport region observes all total 130 sexual crimes in the past 10 years whereas university at buffalo south campus observes all total 33 sexual crimes.